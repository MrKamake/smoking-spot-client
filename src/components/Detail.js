import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import './Detail.scss';

import Rating from 'react-rating';
import jwt from 'jsonwebtoken';

let userToken = localStorage.getItem('jwtToken');
let decodedToken;
if (userToken) {
  decodedToken = jwt.verify(userToken, process.env.REACT_APP_SECRET_KEY);
}

const Detail = ({
  match: { params },
  getSpotDetail,
  spotDetail,
  onSubmit,
  getComents,
  comentList
}) => {
  const [rate, setRate] = useState(0);
  const [comentText, setComentText] = useState('');

  useEffect(() => {
    getSpotDetail(params.id);
    getComents(params.id);
    return () => {};
  }, []);

  const _handleSubmit = ev => {
    ev.preventDefault();
    onSubmit({
      spot_id: params.id,
      created_by: decodedToken.email,
      created_at: new Date(),
      coment_text: comentText,
      rate
    });
  };

  return (
    <>
      <div className='header-box'>
        <Link to='/spots/new' className='new-page-button'>
          <img
            className='new-button-icon'
            alt='new-button'
            src='https://www.iconsdb.com/icons/preview/white/plus-5-xxl.png'
          />
        </Link>
        <Link to='/home' className='sub-logo'>
          <div className='sub-logo-smoking'>SMOKING</div>
          <div className='sub-logo-spot'>SPOT</div>
        </Link>
        <Link to='/spots/map' className='search-button'>
          <img
            className='search-button-icon'
            alt='search box'
            src='https://www.iconsdb.com/icons/preview/white/search-12-xxl.png'
          />
        </Link>
      </div>
      {spotDetail && (
        <div className='spot-detail'>
          <div className='image-N-creator'>
            <img
              className='spot-detail-image'
              src={spotDetail.photo_url}
              alt={spotDetail.title}
            />
            <div className='spot-detail-creator'>{spotDetail.email}</div>
          </div>
          <div className='description-box'>
            <div className='spot-detail-title'>{spotDetail.title}</div>
            <div className='spot-detail-rate'>
              <Rating
                className='rate-box'
                initialRating={
                  comentList.reduce((acc, coment) => acc + coment.rate, 0) /
                  comentList.length
                }
                readonly
                emptySymbol={
                  <img
                    src='https://cdn3.iconfinder.com/data/icons/essentials-volume-i/128/star-3-512.png'
                    className='icon'
                    alt='icon'
                  />
                }
                fullSymbol={
                  <img
                    src='https://cdn1.iconfinder.com/data/icons/vote-reward-7/24/award_reward_rate_rating_star_empty-512.png'
                    className='icon'
                    alt='icon'
                  />
                }
              />
            </div>
            <div className='spot-detail-address'>
              주소: {spotDetail.address}
            </div>
          </div>
          <form className='user-input-box' onSubmit={_handleSubmit}>
            <div className='right-input-box'>
              <input
                className='text-input'
                type='text'
                placeholder='댓글을 남겨주세요..'
                onChange={ev => setComentText(ev.target.value)}
              />
              <Rating
                className='rate-box rate-input'
                onClick={clickedRate => setRate(clickedRate)}
                emptySymbol={
                  <img
                    src='https://cdn3.iconfinder.com/data/icons/essentials-volume-i/128/star-3-512.png'
                    className='icon'
                    alt='icon'
                  />
                }
                fullSymbol={
                  <img
                    src='https://cdn1.iconfinder.com/data/icons/vote-reward-7/24/award_reward_rate_rating_star_empty-512.png'
                    className='icon'
                    alt='icon'
                  />
                }
              />
            </div>
            <input
              className={`user-submit ${!!comentText}`}
              type='submit'
              value='등록'
              disabled={!comentText}
            />
          </form>
          {comentList && (
            <div className='coment-list-box'>
              {comentList.map((coment, index) => (
                <div className='coment-box' key={index}>
                  <div className='create-by'>{coment.created_by}</div>
                  <div className='coment-text'>{coment.coment_text}</div>
                  <div className='created-at'>
                    {coment.created_at.slice(0, 10)}
                  </div>
                  <Rating
                    className='rate-box coment'
                    initialRating={coment.rate}
                    readonly
                    emptySymbol={
                      <img
                        src='https://cdn3.iconfinder.com/data/icons/essentials-volume-i/128/star-3-512.png'
                        className='icon'
                        alt='icon'
                      />
                    }
                    fullSymbol={
                      <img
                        src='https://cdn1.iconfinder.com/data/icons/vote-reward-7/24/award_reward_rate_rating_star_empty-512.png'
                        className='icon'
                        alt='icon'
                      />
                    }
                  />
                </div>
              ))}
            </div>
          )}
        </div>
      )}
    </>
  );
};

export default Detail;
