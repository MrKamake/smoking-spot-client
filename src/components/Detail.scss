.spot-detail {
  text-align: initial;
  .image-N-creator {
    position: relative;
    .spot-detail-image {
      width: 100%;
    }
    .spot-detail-creator {
      position: absolute;
      right: 0;
      bottom: 1%;
      padding: 3px 5px;
      font-size: 0.9rem;
      color: lightgrey;
      background-color: black;
    }
  }
  .description-box {
    position: relative;
    padding: 20px;
    border-bottom: 1px solid;
    .spot-detail-title {
      display: inline-block;
      padding-bottom: 10px;
      font-size: 1.3rem;
      font-weight: bold;
    }
    .spot-detail-rate {
      position: absolute;
      right: 20%;
      display: inline-block;
    }
  }
  .spot-detail-address {
    font-weight: bold;
  }
  .user-input-box {
    max-width: 100%;
    padding: 10px 0;
    border-bottom: 1px solid;
    .right-input-box {
      display: inline-block;
      vertical-align: top;
      padding-left: 10px;
      width: 80%;
      .text-input {
        display: block;
        padding: 5px 10px;
        width: 90%;
        border: 1px solid lightgrey;
        border-radius: 4px;
      }
      .rate-input {
        display: block;
      }
    }
    .user-submit {
      display: inline-block;
      width: 15%;
      height: 100%;
      padding: 10px;
      border: 0;
      border-radius: 4px;
      font-size: 1.1rem;
      font-weight: bold;
      color: white;
      background-color: rgb(80, 170, 121);
    }
    .false {
      background-color: darkgrey;
    }
  }
  .icon {
    width: 20px;
  }
}
