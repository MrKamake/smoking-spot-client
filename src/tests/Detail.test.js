import React from 'react';
import { shallow } from 'enzyme';
import Detail from '../components/Detail';

describe('<Detail />', () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);
  let spotDetail;

  beforeEach(() => {
    wrapper = shallow(<Detail match={'88asdasd7666f771gvav75'} />);
    spotDetail = {
      title: 'Lizard',
      email: 'kamake@google.com',
      address: 'Poketmon world'
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const logoText = wrapper.find('.sub-logo-smoking');
    const logoSubText = wrapper.find('.sub-logo-spot');

    expect(logoText.contains('SMOKING')).toBe(true);
    expect(logoSubText.contains('SPOT')).toBe(true);
  });

  it('renders spotDetail what is one of props', () => {
    wrapper = shallow(
      <Detail
        spotDetail={spotDetail}
        comentList={[]}
        match={'88asdasd7666f771gvav75'}
      />
    );

    const creator = wrapper.find('.spot-detail-creator');
    const title = wrapper.find('.spot-detail-title');
    const address = wrapper.find('.spot-detail-address');

    expect(creator.contains('kamake@google.com')).toBe(true);
    expect(title.contains('Lizard')).toBe(true);
    expect(address.contains('Poketmon world')).toBe(true);
  });

  it('renders comentList what is one of props', () => {
    const comentList = [
      {
        created_by: 'Jiwoo',
        coment_text: 'U r mine from now!!',
        created_at: '2019-10-25-AM'
      }
    ];
    wrapper = shallow(
      <Detail
        spotDetail={spotDetail}
        comentList={comentList}
        match={'88asdasd7666f771gvav75'}
      />
    );

    expect(wrapper.contains(<div className='create-by'>Jiwoo</div>)).toBe(true);
    expect(
      wrapper.contains(<div className='coment-text'>U r mine from now!!</div>)
    ).toBe(true);
    expect(wrapper.contains(<div className='created-at'>2019-10-25</div>)).toBe(
      true
    );
  });
});
