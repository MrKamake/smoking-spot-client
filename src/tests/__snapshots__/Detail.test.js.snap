// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> matches snapshot 1`] = `
<Fragment>
  <div
    className="header-box"
  >
    <Link
      className="new-page-button"
      to="/spots/new"
    >
      <img
        alt="new-button"
        className="new-button-icon"
        src="https://www.iconsdb.com/icons/preview/white/plus-5-xxl.png"
      />
    </Link>
    <Link
      className="sub-logo"
      to="/home"
    >
      <div
        className="sub-logo-smoking"
      >
        SMOKING
      </div>
      <div
        className="sub-logo-spot"
      >
        SPOT
      </div>
    </Link>
    <Link
      className="search-button"
      to="/spots/map"
    >
      <img
        alt="search box"
        className="search-button-icon"
        src="https://www.iconsdb.com/icons/preview/white/search-12-xxl.png"
      />
    </Link>
  </div>
</Fragment>
`;
